File: C#/0095. Unique Binary Search Trees II.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<TreeNode> GenerateTrees(int n) {
        return n==0?new List<TreeNode>():GenerateTrees(1,n+1);
    }
    public IList<TreeNode> GenerateTrees(int l,int r){
        IList<TreeNode> result=new List<TreeNode>();
        if(l==r){
            result.Add(null);
            return result;
        }
        for(int m=l;m<r;m++){
            var lList=GenerateTrees(l,m);
            var rList=GenerateTrees(m+1,r);
            foreach(var lNode in lList){
                foreach(var rNode in rList){
                    var node=new TreeNode(m);
                    node.left=lNode;
                    node.right=rNode;
                    result.Add(node);
                }
            }
        }
        return result;
    }
}

File: archives/leetcode2/second/0095. Unique Binary Search Trees II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None(
#         self.right = None

class Solution:
    def generateTrees(self, n: 'int') -> 'List[TreeNode]':
        if n==0:
            return []
        
        nums=[x for x in range(1,n+1)]
        
        def dfs(nums):
            if not nums:
                return [None]
            rep=[]
            for k in range(len(nums)):
                l_list=dfs(nums[:k])
                r_list=dfs(nums[k+1:])
                for l in l_list:
                    for r in r_list:
                        root=TreeNode(nums[k])
                        root.left=l
                        root.right=r
                        rep.append(root)
            return rep
        
        return dfs(nums)
